core/store/src/trie/ops/iter.rs

```diff
@@ -0,0 +1,421 @@
+                IterStep::Descend(*child)
+            }
+            (CrumbStatus::AtChild(i), GenericTrieNode::Branch { children, .. }) => {
+                if i == 0 {
+                    self.key_nibbles.push(0);
+                }
+
+                if let Some(ref child) = children[i as usize] {
+                    if i != 0 {
+                        *self.key_nibbles.last_mut().expect("Pushed child value before") = i;
+                    }
+                    IterStep::Descend(*child)
+                } else {
+                    IterStep::Continue
+                }
+            }
+            _ => panic!("Should never see Entering or AtChild without a Branch here."),
+        };
+        Some(step)
+    }
+}
+
+#[derive(Debug)]
+enum IterStep<N, V> {
+    Continue,
+    PopTrail,
+    Descend(N),
+    Value(V),
+}
+
+impl<N, V, I> Iterator for TrieIteratorImpl<N, V, I>
+where
+    N: Copy,
+    V: Clone,
+    I: TrieIteratorStorageInterface<N, V>,
+{
+    type Item = Result<TrieItem, StorageError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let iter_step = self.iter_step()?;
+
+            let should_prune = match self.prune_condition {
+                Some(ref prune_condition) => prune_condition(&self.key_nibbles),
+                None => false,
+            };
+
+            match (iter_step, should_prune) {
+                (IterStep::Continue, _) => {}
+                (IterStep::PopTrail, _) => {
+                    self.trail.pop();
+                }
+                // Skip processing node if it should be pruned.
+                (_, true) => {}
+                (IterStep::Descend(ptr), false) => match self.descend_into_node(Some(ptr)) {
+                    Ok(_) => {}
+                    Err(e) => return Some(Err(e)),
+                },
+                (IterStep::Value(value_ref), false) => {
+                    let value = self.trie_interface.get_and_record_value(value_ref);
+                    return Some(value.map(|value| (self.key(), value)));
+                }
+            }
+        }
+    }
+}
+
+/// *************************************************************************************************
+/// Section below is used for state sync. Ideally we should have a better way of reading the trie
+/// nodes and working with the iterator, example using trie recorder and prune conditions but while
+/// we figure that out, we can keep the implementation below
+/// *************************************************************************************************
+
+// Item extracted from Trie during depth first traversal, corresponding to some Trie node.
+#[derive(Debug)]
+pub struct TrieTraversalItem {
+    /// Hash of the node.
+    pub hash: CryptoHash,
+    /// Key of the node if it stores a value.
+    pub key: Option<Vec<u8>>,
+}
+
+// Extension for State Parts processing
+impl<I> TrieIteratorImpl<CryptoHash, ValueHandle, I>
```

Comment:
I could only implement this function for DiskTrieIterator as the type of `Node` and `Value` is the same, i.e. CryptoHash. It doesn't work for a generic implementation to store `TrieTraversalItem`.

Ideally we should get rid of this code in favor of using trie recorder :/